experimental/tools/cpp-simulation/cpp-simulation.cpp

```diff
@@ -0,0 +1,2237 @@
+  const ConsumerDescendants cons;
+};
+
+////--- Control (no data)
+
+/// In this case the user can change the ready signal, but has
+/// ReadOnly access to the valid one.
+struct ControlConsumerRW : public ConsumerRW {
+  ControlConsumerRW(ControlState &reader, ControlState &writer)
+      : ConsumerRW(reader.valid, writer.ready, D_ControlConsumerRW) {}
+
+  static bool classof(const ConsumerRW *c) {
+    return c->getType() == D_ControlConsumerRW;
+  }
+};
+
+/// In this case the user can change the valid signal, but has
+/// ReadOnly access to the ready one.
+struct ControlProducerRW : public ProducerRW {
+  ControlProducerRW(ControlState &reader, ControlState &writer)
+      : ProducerRW(writer.valid, reader.ready, D_ControlProducerRW) {}
+
+  static bool classof(const ProducerRW *c) {
+    return c->getType() == D_ControlProducerRW;
+  }
+};
+
+////--- Channel (data)
+
+/// In this case the user can change the valid signal, but has ReadOnly access
+/// to the valid and data ones.
+struct ChannelConsumerRW : public ConsumerRW {
+  const Data &data;
+
+  ChannelConsumerRW(ChannelState &reader, ChannelState &writer)
+      : ConsumerRW(reader.valid, writer.ready, D_ChannelConsumerRW),
+        data(reader.data) {}
+
+  static bool classof(const ConsumerRW *c) {
+    return c->getType() == D_ChannelConsumerRW;
+  }
+};
+
+/// In this case the user can change the valid and data signals, but has
+/// ReadOnly access to the ready one.
+struct ChannelProducerRW : public ProducerRW {
+  Data &data;
+
+  ChannelProducerRW(ChannelState &reader, ChannelState &writer)
+      : ProducerRW(writer.valid, reader.ready, D_ChannelProducerRW),
+        data(writer.data) {}
+
+  static bool classof(const ProducerRW *c) {
+    return c->getType() == D_ChannelProducerRW;
+  }
+};
+
+//===----------------------------------------------------------------------===//
+// Readers & writers
+//===----------------------------------------------------------------------===//
+/// The following structures are used to represent components that can be
+/// either datafull or dataless. They store a pointer to Data and some
+/// user-friendly functions.
+
+/// A struct to represent consumer's data
+struct ConsumerData {
+  Data const *data = nullptr;
+  unsigned dataWidth = 0;
+
+  ConsumerData(ConsumerRW *ins) {
+    if (auto *p = dyn_cast<ChannelConsumerRW>(ins)) {
+      data = &p->data;
+      dataWidth = data->bitwidth;
+    } else
+      data = nullptr;
+  }
+
+  bool hasValue() const { return data != nullptr; }
+};
+
+/// A struct to represent producer's data
+struct ProducerData {
+  Data *data = nullptr;
+  unsigned dataWidth = 0;
+
+  ProducerData(ProducerRW *outs) {
+    if (auto *p = dyn_cast<ChannelProducerRW>(outs)) {
+      data = &p->data;
+    } else
+      data = nullptr;
+  }
+
+  ProducerData &operator=(const ConsumerData &other) {
+    if (data)
+      *data = *other.data;
+    return *this;
+  }
+
+  bool hasValue() const { return data != nullptr; }
+};
```

Comment:
I see that's your solution to handling components that can be dataless or dataful. That seems to work fine and yield readable code, great!